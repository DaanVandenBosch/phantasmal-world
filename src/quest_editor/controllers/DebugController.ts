import { Controller } from "../../core/controllers/Controller";
import { Property } from "../../core/observable/property/Property";
import { ListProperty } from "../../core/observable/property/list/ListProperty";
import { QuestEditorStore } from "../stores/QuestEditorStore";
import { GuiStore, GuiTool } from "../../core/stores/GuiStore";
import { LogEntry } from "../../core/Logger";
import { LogStore } from "../stores/LogStore";
import { Severity } from "../../core/Severity";
import { property } from "../../core/observable";

export class DebugController extends Controller {
    readonly can_debug: Property<boolean>;
    readonly can_step: Property<boolean>;
    readonly can_stop: Property<boolean>;
    readonly thread_ids: ListProperty<number>;
    readonly debugging_thread_id: Property<number | undefined>;
    readonly active_thread_id: Property<number | undefined>;
    readonly can_select_thread: Property<boolean>;
    readonly log: ListProperty<LogEntry>;
    readonly severity: Property<Severity>;

    constructor(
        gui_store: GuiStore,
        private readonly quest_editor_store: QuestEditorStore,
        private readonly log_store: LogStore,
    ) {
        super();

        this.can_debug = quest_editor_store.current_quest.map(q => q != undefined);

        this.can_step = quest_editor_store.quest_runner.paused;

        this.can_stop = quest_editor_store.quest_runner.running;

        this.thread_ids = quest_editor_store.quest_runner.thread_ids;

        this.debugging_thread_id = quest_editor_store.quest_runner.debugging_thread_id;

        this.active_thread_id = quest_editor_store.quest_runner.active_thread_id;

        this.can_select_thread = quest_editor_store.quest_runner.thread_ids.map(
            ids => ids.length > 0 && quest_editor_store.quest_runner.running.val,
        );

        this.log = log_store.log;
        this.severity = log_store.severity;

        this.disposables(
            gui_store.on_global_keydown(GuiTool.QuestEditor, "F5", this.debug),

            gui_store.on_global_keydown(GuiTool.QuestEditor, "Shift-F5", this.stop),

            gui_store.on_global_keydown(GuiTool.QuestEditor, "F6", this.resume),

            gui_store.on_global_keydown(GuiTool.QuestEditor, "F8", this.step_over),

            gui_store.on_global_keydown(GuiTool.QuestEditor, "F7", this.step_in),

            gui_store.on_global_keydown(GuiTool.QuestEditor, "Shift-F8", this.step_out),
        );
    }

    debug = (): void => {
        const quest = this.quest_editor_store.current_quest.val;

        if (quest) {
            this.quest_editor_store.quest_runner.run(quest);
        }
    };

    resume = (): void => {
        this.quest_editor_store.quest_runner.resume();
    };

    step_over = (): void => {
        this.quest_editor_store.quest_runner.step_over();
    };

    step_in = (): void => {
        this.quest_editor_store.quest_runner.step_into();
    };

    step_out = (): void => {
        this.quest_editor_store.quest_runner.step_out();
    };

    stop = (): void => {
        this.quest_editor_store.quest_runner.stop();
    };

    set_severity = (severity: Severity): void => {
        this.log_store.set_severity(severity);
    };

    select_thread = (thread_id: number): void => {
        this.quest_editor_store.quest_runner.set_debugging_thread(thread_id);
    };
}
