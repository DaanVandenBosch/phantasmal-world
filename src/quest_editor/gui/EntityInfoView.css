.quest_editor_EntityInfoView {
    outline: none;
    box-sizing: border-box;
    padding: 3px;
    overflow: auto;
}

.quest_editor_EntityInfoView table {
    table-layout: fixed;
    width: 100%;
    margin: 0 auto;
}

.quest_editor_EntityInfoView th {
    text-align: left;
}

.quest_editor_EntityInfoView th.quest_editor_EntityInfoView_coord {
    padding-left: 10px;
}

.quest_editor_EntityInfoView .core_NumberInput {
    width: 100%;
}
