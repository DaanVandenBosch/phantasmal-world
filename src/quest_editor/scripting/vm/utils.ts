export type BinaryNumericOperation = (a: number, b: number) => number;

export const numeric_ops: Record<
    "add" | "sub" | "mul" | "div" | "idiv" | "mod" | "and" | "or" | "xor" | "shl" | "shr",
    BinaryNumericOperation
> = {
    add: (a, b) => a + b,
    sub: (a, b) => a - b,
    mul: (a, b) => a * b,
    div: (a, b) => a / b,
    idiv: (a, b) => Math.floor(a / b),
    mod: (a, b) => a % b,
    and: (a, b) => a & b,
    or: (a, b) => a | b,
    xor: (a, b) => a ^ b,
    shl: (a, b) => a << b,
    shr: (a, b) => a >>> b,
};

export type ComparisonOperation = (a: number, b: number) => boolean;

/**
 * Short-circuiting fold.
 */
export function andfold<T, A>(
    fn: (acc: A, cur: T) => A | undefined,
    init: A,
    lst: T[],
): A | undefined {
    let acc = init;

    for (const item of lst) {
        const new_val = fn(acc, item);

        if (new_val === undefined) {
            return undefined;
        } else {
            acc = new_val;
        }
    }

    return acc;
}

/**
 * Short-circuiting reduce.
 */
export function andreduce<T>(fn: (acc: T, cur: T) => T | undefined, lst: T[]): T | undefined {
    return andfold(fn, lst[0], lst.slice(1));
}

/**
 * Applies the given arguments to the given function.
 * Returns the second argument if the function returns a truthy value, else undefined.
 */
export function andsecond<T>(fn: (first: T, second: T) => any, first: T, second: T): T | undefined {
    if (fn(first, second)) {
        return second;
    }
    return undefined;
}

export function rest<T>(lst: T[]): T[] {
    return lst.slice(1);
}
