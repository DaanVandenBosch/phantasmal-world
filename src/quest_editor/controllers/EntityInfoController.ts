import { Controller } from "../../core/controllers/Controller";
import { QuestEditorStore } from "../stores/QuestEditorStore";
import { Property } from "../../core/observable/property/Property";
import { QuestNpcModel } from "../model/QuestNpcModel";
import { flat_map_to_list, list_property, property } from "../../core/observable";
import { Euler, Vector3 } from "three";
import { deg_to_rad } from "../../core/math";
import { TranslateEntityAction } from "../actions/TranslateEntityAction";
import { RotateEntityAction } from "../actions/RotateEntityAction";
import { euler } from "../model/euler";
import { entity_data } from "../../core/data_formats/parsing/quest/Quest";
import { ListProperty } from "../../core/observable/property/list/ListProperty";
import { QuestEntityPropModel } from "../model/QuestEntityPropModel";

const DUMMY_VECTOR = Object.freeze(new Vector3());
const DUMMY_EULER = Object.freeze(new Euler());

export class EntityInfoController extends Controller {
    readonly unavailable: Property<boolean>;
    readonly enabled: Property<boolean>;
    readonly type: Property<string>;
    readonly name: Property<string>;
    readonly section_id: Property<string>;
    readonly wave: Property<string>;
    readonly wave_hidden: Property<boolean>;
    readonly position: Property<Vector3>;
    readonly rotation: Property<Euler>;
    readonly props: ListProperty<QuestEntityPropModel>;

    constructor(private readonly store: QuestEditorStore) {
        super();

        const entity = store.selected_entity;
        this.unavailable = entity.map(e => e == undefined);
        this.enabled = store.quest_runner.running.map(r => !r);

        this.type = entity.map(e => (e instanceof QuestNpcModel ? "NPC" : "Object"));
        this.name = entity.map(e => (e == undefined ? "" : entity_data(e.type).name));
        this.section_id = entity.flat_map(e =>
            e == undefined ? property("") : e.section_id.map(id => id.toString()),
        );
        this.wave = entity.flat_map(e =>
            e instanceof QuestNpcModel
                ? e.wave.flat_map(w => w?.id?.map(w => w.toString()) ?? property("None"))
                : property(""),
        );
        this.wave_hidden = entity.map(e => !(e instanceof QuestNpcModel));
        this.position = entity.flat_map(e => e?.position ?? property(DUMMY_VECTOR));
        this.rotation = entity.flat_map(e => e?.rotation ?? property(DUMMY_EULER));
        this.props = flat_map_to_list(e => e?.props ?? list_property(), entity);
    }

    focused = (): void => {
        this.store.undo.make_current();
    };

    set_pos_x(x: number): void {
        const entity = this.store.selected_entity.val;

        if (entity) {
            const pos = entity.position.val;
            this.store.undo
                .push(
                    new TranslateEntityAction(
                        this.store,
                        entity,
                        entity.section.val,
                        entity.section.val,
                        pos,
                        new Vector3(x, pos.y, pos.z),
                        false,
                    ),
                )
                .redo();
        }
    }

    set_pos_y(y: number): void {
        const entity = this.store.selected_entity.val;

        if (entity) {
            const pos = entity.position.val;
            this.store.undo
                .push(
                    new TranslateEntityAction(
                        this.store,
                        entity,
                        entity.section.val,
                        entity.section.val,
                        pos,
                        new Vector3(pos.x, y, pos.z),
                        false,
                    ),
                )
                .redo();
        }
    }

    set_pos_z(z: number): void {
        const entity = this.store.selected_entity.val;

        if (entity) {
            const pos = entity.position.val;
            this.store.undo
                .push(
                    new TranslateEntityAction(
                        this.store,
                        entity,
                        entity.section.val,
                        entity.section.val,
                        pos,
                        new Vector3(pos.x, pos.y, z),
                        false,
                    ),
                )
                .redo();
        }
    }

    set_rot_x(x: number): void {
        const entity = this.store.selected_entity.val;

        if (entity) {
            const rot = entity.rotation.val;
            this.store.undo
                .push(
                    new RotateEntityAction(
                        this.store,
                        entity,
                        rot,
                        euler(deg_to_rad(x), rot.y, rot.z),
                        false,
                    ),
                )
                .redo();
        }
    }

    set_rot_y(y: number): void {
        const entity = this.store.selected_entity.val;

        if (entity) {
            const rot = entity.rotation.val;
            this.store.undo
                .push(
                    new RotateEntityAction(
                        this.store,
                        entity,
                        rot,
                        euler(rot.x, deg_to_rad(y), rot.z),
                        false,
                    ),
                )
                .redo();
        }
    }

    set_rot_z(z: number): void {
        const entity = this.store.selected_entity.val;

        if (entity) {
            const rot = entity.rotation.val;
            this.store.undo
                .push(
                    new RotateEntityAction(
                        this.store,
                        entity,
                        rot,
                        euler(rot.x, rot.y, deg_to_rad(z)),
                        false,
                    ),
                )
                .redo();
        }
    }
}
