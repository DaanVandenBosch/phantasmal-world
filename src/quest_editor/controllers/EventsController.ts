import { Controller } from "../../core/controllers/Controller";
import { QuestEditorStore } from "../stores/QuestEditorStore";
import { Property } from "../../core/observable/property/Property";
import { QuestEventDagModel } from "../model/QuestEventDagModel";
import { ListProperty } from "../../core/observable/property/list/ListProperty";
import { flat_map_to_list, list_property, map } from "../../core/observable";
import { QuestEventModel } from "../model/QuestEventModel";
import { EditEventDelayAction } from "../actions/EditEventDelayAction";
import { WaveModel } from "../model/WaveModel";
import { RemoveEventAction } from "../actions/RemoveEventAction";
import { CreateEventAction } from "../actions/CreateEventAction";

export class EventsController extends Controller {
    readonly event_dag: Property<QuestEventDagModel | undefined>;
    readonly event_sub_graphs: ListProperty<ListProperty<QuestEventModel>>;
    readonly enabled: Property<boolean>;
    readonly unavailable: Property<boolean>;

    constructor(private readonly store: QuestEditorStore) {
        super();

        this.enabled = store.quest_runner.running.map(r => !r);
        this.unavailable = store.current_quest.map(q => q == undefined);

        this.event_dag = map(
            (quest, area) => {
                if (quest && area) {
                    return quest.event_dags.get(area.id);
                } else {
                    return undefined;
                }
            },
            store.current_quest,
            store.current_area,
        );

        this.event_sub_graphs = flat_map_to_list(
            dag => dag?.connected_sub_graphs ?? list_property(),
            this.event_dag,
        );
    }

    focused = (): void => {
        this.store.undo.make_current();
    };

    is_selected(event: QuestEventModel): Property<boolean> {
        return this.store.selected_wave.map(selected => event.wave === selected);
    }

    add_event = (): void => {
        const quest = this.store.current_quest.val;
        const area = this.store.current_area.val;

        if (quest && area) {
            const event_dag = quest.get_event_dag_or_create(area.id);

            const parent = event_dag.events.find(
                event => event.wave === this.store.selected_wave.val,
            );

            const section_id =
                parent?.section_id ?? this.store.selected_entity.val?.section_id?.val ?? 1;

            const event_ids: number[] = [];
            const wave_ids: number[] = [];

            for (const event of event_dag.events) {
                event_ids.push(event.id);

                if (event.wave.section_id.val === section_id) {
                    wave_ids.push(event.wave.id.val);
                }
            }

            event_ids.sort((a, b) => a - b);
            wave_ids.sort((a, b) => a - b);

            // Find the first available wave id.
            let wave_id: number = 0;

            for (const existing_wave_id of wave_ids) {
                if (++wave_id !== existing_wave_id) {
                    break;
                }
            }

            if (wave_id === wave_ids.length) {
                wave_id++;
            }

            // Generate id.
            let id_str: string;

            if (parent) {
                // Generate id based on first ancestor id and amount of its child events.
                let ancestor = parent;

                while (true) {
                    // Always choose the first parent.
                    const [p] = event_dag.get_parents(ancestor);

                    if (p) {
                        ancestor = p;
                    } else {
                        break;
                    }
                }

                const sub_graph_size = event_dag.get_sub_graph(ancestor).length.val;
                id_str = `${ancestor.id}${sub_graph_size}`;
            } else {
                // Generate id based on section id and wave id.
                id_str = `${section_id}${wave_id}`;
            }

            let id = parseInt(id_str, 10);

            // Make sure id is unique.
            let existing_index: number = 0;

            while (true) {
                existing_index = event_ids.indexOf(id, existing_index);

                if (existing_index === -1) {
                    break;
                } else {
                    id++;
                }
            }

            this.store.undo
                .push(
                    new CreateEventAction(
                        quest,
                        event_dag,
                        new QuestEventModel(
                            id,
                            section_id,
                            new WaveModel(wave_id, area.id, section_id),
                            30,
                            0, // TODO: what is a sensible value for event.unknown?
                        ),
                        parent,
                    ),
                )
                .redo();
        }
    };

    remove_event = (event: QuestEventModel): void => {
        const quest = this.store.current_quest.val;
        const dag = this.event_dag.val;

        if (quest && dag) {
            this.store.undo.push(new RemoveEventAction(this.store, quest, dag, event)).redo();
        }
    };

    set_selected_wave = (wave?: WaveModel): void => {
        if (this.enabled.val) {
            this.store.set_selected_wave(wave);
        }
    };

    set_delay = (event: QuestEventModel, delay: number): void => {
        this.store.undo.push(new EditEventDelayAction(event, event.delay.val, delay)).redo();
    };
}
