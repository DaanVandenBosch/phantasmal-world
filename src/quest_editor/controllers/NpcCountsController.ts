import { Controller } from "../../core/controllers/Controller";
import { QuestEditorStore } from "../stores/QuestEditorStore";
import { Property } from "../../core/observable/property/Property";
import { QuestNpcModel } from "../model/QuestNpcModel";
import { npc_data, NpcType } from "../../core/data_formats/parsing/quest/npc_types";
import { property } from "../../core/observable";

export type NameWithCount = { readonly name: string; readonly count: number };

export class NpcCountsController extends Controller {
    readonly npc_counts: Property<readonly NameWithCount[]>;
    readonly unavailable: Property<boolean>;

    constructor(store: QuestEditorStore) {
        super();

        this.unavailable = store.current_quest.map(q => q == undefined);

        this.npc_counts = store.current_quest
            .flat_map(quest => (quest ? quest.npcs : property([])))
            .map(this.update_view);
    }

    private update_view(npcs: readonly QuestNpcModel[]): NameWithCount[] {
        const npc_counts = new Map<NpcType, number>();

        for (const npc of npcs) {
            const val = npc_counts.get(npc.type) || 0;
            npc_counts.set(npc.type, val + 1);
        }

        const extra_canadines = (npc_counts.get(NpcType.Canane) || 0) * 8;

        // Sort by canonical order.
        const sorted_npc_counts = [...npc_counts].sort((a, b) => a[0] - b[0]);

        return sorted_npc_counts.map(([npc_type, count]) => {
            const extra = npc_type === NpcType.Canadine ? extra_canadines : 0;

            return { name: npc_data(npc_type).name, count: count + extra };
        });
    }
}
