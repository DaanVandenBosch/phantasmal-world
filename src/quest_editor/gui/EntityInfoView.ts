import { ResizableWidget } from "../../core/gui/ResizableWidget";
import { el } from "../../core/gui/dom";
import { DisabledView } from "./DisabledView";
import { quest_editor_store } from "../stores/QuestEditorStore";
import { QuestNpcModel } from "../model/QuestNpcModel";
import { entity_data } from "../../core/data_formats/parsing/quest/entities";
import "./EntityInfoView.css";
import { NumberInput } from "../../core/gui/NumberInput";
import { Disposer } from "../../core/observable/Disposer";
import { Property } from "../../core/observable/property/Property";
import { Vec3 } from "../../core/data_formats/vector";
import { QuestEntityModel } from "../model/QuestEntityModel";

export class EntityInfoView extends ResizableWidget {
    readonly element = el.div({ class: "quest_editor_EntityInfoView", tab_index: -1 });

    private readonly no_entity_view = new DisabledView("No entity selected.");

    private readonly table_element = el.table();

    private readonly type_element: HTMLTableCellElement;
    private readonly name_element: HTMLTableCellElement;
    private readonly section_id_element: HTMLTableCellElement;
    private readonly pos_x_element = this.disposable(
        new NumberInput(0, { width: 80, round_to: 3 }),
    );
    private readonly pos_y_element = this.disposable(
        new NumberInput(0, { width: 80, round_to: 3 }),
    );
    private readonly pos_z_element = this.disposable(
        new NumberInput(0, { width: 80, round_to: 3 }),
    );
    private readonly world_pos_x_element = this.disposable(
        new NumberInput(0, { width: 80, round_to: 3 }),
    );
    private readonly world_pos_y_element = this.disposable(
        new NumberInput(0, { width: 80, round_to: 3 }),
    );
    private readonly world_pos_z_element = this.disposable(
        new NumberInput(0, { width: 80, round_to: 3 }),
    );

    private readonly entity_disposer = new Disposer();

    constructor() {
        super();

        const entity = quest_editor_store.selected_entity;
        const no_entity = entity.map(e => e == undefined);
        const coord_class = "quest_editor_EntityInfoView_coord";

        this.table_element.append(
            el.tr({}, el.th({ text: "Type:" }), (this.type_element = el.td())),
            el.tr({}, el.th({ text: "Name:" }), (this.name_element = el.td())),
            el.tr({}, el.th({ text: "Section:" }), (this.section_id_element = el.td())),
            el.tr({}, el.th({ text: "Section position:", col_span: 2 })),
            el.tr(
                {},
                el.th({ text: "X:", class: coord_class }),
                el.td({}, this.pos_x_element.element),
            ),
            el.tr(
                {},
                el.th({ text: "Y:", class: coord_class }),
                el.td({}, this.pos_y_element.element),
            ),
            el.tr(
                {},
                el.th({ text: "Z:", class: coord_class }),
                el.td({}, this.pos_z_element.element),
            ),
            el.tr({}, el.th({ text: "World position:", col_span: 2 })),
            el.tr(
                {},
                el.th({ text: "X:", class: coord_class }),
                el.td({}, this.world_pos_x_element.element),
            ),
            el.tr(
                {},
                el.th({ text: "Y:", class: coord_class }),
                el.td({}, this.world_pos_y_element.element),
            ),
            el.tr(
                {},
                el.th({ text: "Z:", class: coord_class }),
                el.td({}, this.world_pos_z_element.element),
            ),
        );

        this.element.append(this.table_element, this.no_entity_view.element);

        this.element.addEventListener("focus", () => quest_editor_store.undo.make_current(), true);

        this.bind_hidden(this.table_element, no_entity);

        this.disposables(
            this.no_entity_view.visible.bind_to(no_entity),

            entity.observe(({ value: entity }) => {
                this.entity_disposer.dispose_all();

                if (entity) {
                    this.type_element.innerText =
                        entity instanceof QuestNpcModel ? "NPC" : "Object";
                    const name = entity_data(entity.type).name;
                    this.name_element.innerText = name;
                    this.name_element.title = name;

                    this.entity_disposer.add(
                        entity.section_id.observe(
                            ({ value: section_id }) => {
                                this.section_id_element.innerText = section_id.toString();
                            },
                            { call_now: true },
                        ),
                    );

                    this.observe(
                        entity,
                        entity.position,
                        false,
                        this.pos_x_element,
                        this.pos_y_element,
                        this.pos_z_element,
                    );

                    this.observe(
                        entity,
                        entity.world_position,
                        true,
                        this.world_pos_x_element,
                        this.world_pos_y_element,
                        this.world_pos_z_element,
                    );
                }
            }),
        );

        this.finalize_construction(EntityInfoView.prototype);
    }

    dispose(): void {
        super.dispose();
        this.entity_disposer.dispose();
    }

    private observe(
        entity: QuestEntityModel,
        pos: Property<Vec3>,
        world: boolean,
        x_input: NumberInput,
        y_input: NumberInput,
        z_input: NumberInput,
    ): void {
        this.entity_disposer.add_all(
            pos.observe(
                ({ value: { x, y, z } }) => {
                    x_input.value.val = x;
                    y_input.value.val = y;
                    z_input.value.val = z;
                },
                { call_now: true },
            ),

            x_input.value.observe(({ value }) =>
                quest_editor_store.translate_entity(
                    entity,
                    entity.section.val,
                    entity.section.val,
                    pos.val,
                    new Vec3(value, pos.val.y, pos.val.z),
                    world,
                ),
            ),

            y_input.value.observe(({ value }) =>
                quest_editor_store.translate_entity(
                    entity,
                    entity.section.val,
                    entity.section.val,
                    pos.val,
                    new Vec3(pos.val.x, value, pos.val.z),
                    world,
                ),
            ),

            z_input.value.observe(({ value }) =>
                quest_editor_store.translate_entity(
                    entity,
                    entity.section.val,
                    entity.section.val,
                    pos.val,
                    new Vec3(pos.val.x, pos.val.y, value),
                    world,
                ),
            ),
        );
    }
}
