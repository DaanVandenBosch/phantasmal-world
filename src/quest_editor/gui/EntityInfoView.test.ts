import { EntityInfoView } from "./EntityInfoView";
import { EntityInfoController } from "../controllers/EntityInfoController";
import {
    create_area_store,
    create_quest_editor_store,
} from "../../../test/src/quest_editor/stores/store_creation";
import { undo_manager } from "../../core/undo/UndoManager";
import { load_default_quest_model, pw_test } from "../../../test/src/utils";

test(
    "Renders correctly without an entity selected.",
    pw_test({}, disposer => {
        const area_store = create_area_store(disposer);
        const store = create_quest_editor_store(disposer, area_store);
        const view = disposer.add(
            new EntityInfoView(disposer.add(new EntityInfoController(store))),
        );

        expect(view.element).toMatchSnapshot('should render a "No entity selected." view');

        store.set_current_quest(load_default_quest_model(area_store));

        expect(view.element).toMatchSnapshot('should render a "No entity selected." view');
    }),
);

test(
    "Renders correctly with an entity selected.",
    pw_test({}, disposer => {
        const area_store = create_area_store(disposer);
        const store = create_quest_editor_store(disposer, area_store);
        const view = disposer.add(
            new EntityInfoView(disposer.add(new EntityInfoController(store))),
        );

        const quest = load_default_quest_model(area_store);
        store.set_current_quest(quest);
        store.set_selected_entity(quest.npcs.get(0));

        expect(view.element).toMatchSnapshot("should render a table of editable properties");
    }),
);

test(
    "When the view's element is focused the quest editor store's undo stack should become the current stack.",
    pw_test({}, disposer => {
        const store = create_quest_editor_store(disposer);
        const view = disposer.add(
            new EntityInfoView(disposer.add(new EntityInfoController(store))),
        );

        // Append view element to DOM to make it focusable.
        document.body.append(view.element);

        undo_manager.make_noop_current();

        view.element.focus();

        expect(undo_manager.current.val).toBe(store.undo);
    }),
);
