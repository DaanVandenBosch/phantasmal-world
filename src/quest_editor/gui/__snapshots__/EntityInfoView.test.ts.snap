// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders correctly with an entity selected.: should render a table of editable properties 1`] = `
<div
  class="quest_editor_EntityInfoView"
  tabindex="-1"
>
  <table>
    <tr>
      <th>
        Type:
      </th>
      <td>
        NPC
      </td>
    </tr>
    <tr>
      <th>
        Name:
      </th>
      <td>
        Guild Lady
      </td>
    </tr>
    <tr>
      <th>
        Section:
      </th>
      <td>
        10
      </td>
    </tr>
    <tr>
      <th>
        Wave:
      </th>
      <td>
        0
      </td>
    </tr>
    <tr>
      <th
        colspan="2"
      >
        Position:
      </th>
    </tr>
    <tr>
      <th
        class="quest_editor_EntityInfoView_coord"
      >
        X:
      </th>
      <td>
        <span
          class="core_NumberInput core_Input"
        >
          <input
            class="core_NumberInput_inner core_Input_inner"
            step="any"
            type="number"
          />
        </span>
      </td>
    </tr>
    <tr>
      <th
        class="quest_editor_EntityInfoView_coord"
      >
        Y:
      </th>
      <td>
        <span
          class="core_NumberInput core_Input"
        >
          <input
            class="core_NumberInput_inner core_Input_inner"
            step="any"
            type="number"
          />
        </span>
      </td>
    </tr>
    <tr>
      <th
        class="quest_editor_EntityInfoView_coord"
      >
        Z:
      </th>
      <td>
        <span
          class="core_NumberInput core_Input"
        >
          <input
            class="core_NumberInput_inner core_Input_inner"
            step="any"
            type="number"
          />
        </span>
      </td>
    </tr>
    <tr>
      <th
        colspan="2"
      >
        Rotation:
      </th>
    </tr>
    <tr>
      <th
        class="quest_editor_EntityInfoView_coord"
      >
        X:
      </th>
      <td>
        <span
          class="core_NumberInput core_Input"
        >
          <input
            class="core_NumberInput_inner core_Input_inner"
            step="any"
            type="number"
          />
        </span>
      </td>
    </tr>
    <tr>
      <th
        class="quest_editor_EntityInfoView_coord"
      >
        Y:
      </th>
      <td>
        <span
          class="core_NumberInput core_Input"
        >
          <input
            class="core_NumberInput_inner core_Input_inner"
            step="any"
            type="number"
          />
        </span>
      </td>
    </tr>
    <tr>
      <th
        class="quest_editor_EntityInfoView_coord"
      >
        Z:
      </th>
      <td>
        <span
          class="core_NumberInput core_Input"
        >
          <input
            class="core_NumberInput_inner core_Input_inner"
            step="any"
            type="number"
          />
        </span>
      </td>
    </tr>
  </table>
  <div
    class="quest_editor_UnavailableView"
    hidden=""
  >
    <label
      class="core_Label disabled"
    >
      No entity selected.
    </label>
  </div>
</div>
`;

exports[`Renders correctly without an entity selected.: should render a "No entity selected." view 1`] = `
<div
  class="quest_editor_EntityInfoView"
  tabindex="-1"
>
  <table
    hidden=""
  >
    <tr>
      <th>
        Type:
      </th>
      <td>
        Object
      </td>
    </tr>
    <tr>
      <th>
        Name:
      </th>
      <td />
    </tr>
    <tr>
      <th>
        Section:
      </th>
      <td />
    </tr>
    <tr
      hidden=""
    >
      <th>
        Wave:
      </th>
      <td />
    </tr>
    <tr>
      <th
        colspan="2"
      >
        Position:
      </th>
    </tr>
    <tr>
      <th
        class="quest_editor_EntityInfoView_coord"
      >
        X:
      </th>
      <td>
        <span
          class="core_NumberInput core_Input"
        >
          <input
            class="core_NumberInput_inner core_Input_inner"
            step="any"
            type="number"
          />
        </span>
      </td>
    </tr>
    <tr>
      <th
        class="quest_editor_EntityInfoView_coord"
      >
        Y:
      </th>
      <td>
        <span
          class="core_NumberInput core_Input"
        >
          <input
            class="core_NumberInput_inner core_Input_inner"
            step="any"
            type="number"
          />
        </span>
      </td>
    </tr>
    <tr>
      <th
        class="quest_editor_EntityInfoView_coord"
      >
        Z:
      </th>
      <td>
        <span
          class="core_NumberInput core_Input"
        >
          <input
            class="core_NumberInput_inner core_Input_inner"
            step="any"
            type="number"
          />
        </span>
      </td>
    </tr>
    <tr>
      <th
        colspan="2"
      >
        Rotation:
      </th>
    </tr>
    <tr>
      <th
        class="quest_editor_EntityInfoView_coord"
      >
        X:
      </th>
      <td>
        <span
          class="core_NumberInput core_Input"
        >
          <input
            class="core_NumberInput_inner core_Input_inner"
            step="any"
            type="number"
          />
        </span>
      </td>
    </tr>
    <tr>
      <th
        class="quest_editor_EntityInfoView_coord"
      >
        Y:
      </th>
      <td>
        <span
          class="core_NumberInput core_Input"
        >
          <input
            class="core_NumberInput_inner core_Input_inner"
            step="any"
            type="number"
          />
        </span>
      </td>
    </tr>
    <tr>
      <th
        class="quest_editor_EntityInfoView_coord"
      >
        Z:
      </th>
      <td>
        <span
          class="core_NumberInput core_Input"
        >
          <input
            class="core_NumberInput_inner core_Input_inner"
            step="any"
            type="number"
          />
        </span>
      </td>
    </tr>
  </table>
  <div
    class="quest_editor_UnavailableView"
  >
    <label
      class="core_Label disabled"
    >
      No entity selected.
    </label>
  </div>
</div>
`;

exports[`Renders correctly without an entity selected.: should render a "No entity selected." view 2`] = `
<div
  class="quest_editor_EntityInfoView"
  tabindex="-1"
>
  <table
    hidden=""
  >
    <tr>
      <th>
        Type:
      </th>
      <td>
        Object
      </td>
    </tr>
    <tr>
      <th>
        Name:
      </th>
      <td />
    </tr>
    <tr>
      <th>
        Section:
      </th>
      <td />
    </tr>
    <tr
      hidden=""
    >
      <th>
        Wave:
      </th>
      <td />
    </tr>
    <tr>
      <th
        colspan="2"
      >
        Position:
      </th>
    </tr>
    <tr>
      <th
        class="quest_editor_EntityInfoView_coord"
      >
        X:
      </th>
      <td>
        <span
          class="core_NumberInput core_Input"
        >
          <input
            class="core_NumberInput_inner core_Input_inner"
            step="any"
            type="number"
          />
        </span>
      </td>
    </tr>
    <tr>
      <th
        class="quest_editor_EntityInfoView_coord"
      >
        Y:
      </th>
      <td>
        <span
          class="core_NumberInput core_Input"
        >
          <input
            class="core_NumberInput_inner core_Input_inner"
            step="any"
            type="number"
          />
        </span>
      </td>
    </tr>
    <tr>
      <th
        class="quest_editor_EntityInfoView_coord"
      >
        Z:
      </th>
      <td>
        <span
          class="core_NumberInput core_Input"
        >
          <input
            class="core_NumberInput_inner core_Input_inner"
            step="any"
            type="number"
          />
        </span>
      </td>
    </tr>
    <tr>
      <th
        colspan="2"
      >
        Rotation:
      </th>
    </tr>
    <tr>
      <th
        class="quest_editor_EntityInfoView_coord"
      >
        X:
      </th>
      <td>
        <span
          class="core_NumberInput core_Input"
        >
          <input
            class="core_NumberInput_inner core_Input_inner"
            step="any"
            type="number"
          />
        </span>
      </td>
    </tr>
    <tr>
      <th
        class="quest_editor_EntityInfoView_coord"
      >
        Y:
      </th>
      <td>
        <span
          class="core_NumberInput core_Input"
        >
          <input
            class="core_NumberInput_inner core_Input_inner"
            step="any"
            type="number"
          />
        </span>
      </td>
    </tr>
    <tr>
      <th
        class="quest_editor_EntityInfoView_coord"
      >
        Z:
      </th>
      <td>
        <span
          class="core_NumberInput core_Input"
        >
          <input
            class="core_NumberInput_inner core_Input_inner"
            step="any"
            type="number"
          />
        </span>
      </td>
    </tr>
  </table>
  <div
    class="quest_editor_UnavailableView"
  >
    <label
      class="core_Label disabled"
    >
      No entity selected.
    </label>
  </div>
</div>
`;
